### ISSUE TITLE
Promise-based queue with limited number of "workers"

### ISSUE THREAD

**Issue by shergin** [mxcl/PromiseKit#444]
I need something like npm-library `promise-queue` (https://www.npmjs.com/package/promise-queue) but in Swift and for PromiseKit.
So, is there any chance that PromiseKit will contain something similar?

**Comment by mxcl** [comment]
Synopsis of what you want? Link doesn't really explain itself.

**Comment by shergin** [comment]
Let's assume that we have a number of items (or stream of items) that we have to process. And we have promise-fabric that can process this type of items.
Usually we can use `when()` function, like this
```
func processor(item: Item) -> Promise<ProcessedItem> {
    return Promise { ... }
}

let items: [Item] = [....]

when(items.map { processor($0) }).then { processedItems in
    // Done!
}
```

That really great but... I know that I have system resources that allow me process only 3 items in parallel (otherwise iOS will kill my app under memory pressure).

So I decided to use pattern that very common (kind of) in front-end-promise-world: promise queue.

It can be done like this:

```
func processor(item: Item) -> Promise<ProcessedItem> {
    return Promise { ... }
}

let items: [Item] = [....]

let queue = PromiseQueue<Item, ProcessedItem>(maxConcurrent: 5, maxQueue: 100)
for item in items {
    queue.add(processor(item))
}

queue.resultPromise.then { processedItems in
   // Done!!
}
```

Disclaimer: I am aware of how GCD's queues work, I believe it cannot be done thought this kind of queues.

My question actually is: Dear mxcl, as maitainter, do you think that such kind of (more high-level) functionality has to be in (pretty low-level, which I really love) library like PromiseKit or not?

PS: This pattern also is very useful with UI/UX programming when we need to deal with some user input but we can throttle some of pieces of user input.

**Comment by mxcl** [comment]
Makes sense. We should have this.

**Comment by mxcl** [comment]
Seems like some parameters to when may be logical? Or would this just be non-intuitive to people experienced with Promises?

```swift
when(promises, concurrently: 5).then {
    //…
}
```

The maxQueue parameter seems a little esoteric.

**Comment by shergin** [comment]
I believe, we actually cannot use/extend `when` because it accepts promises, so if we already have the promises, it immediately means that they already started resolving process (but we want to postpone it).

`maxQueue` parameter can be useful in UI programming when we are implementing something like asynchronous throttling. I have used it.
Let's assume that we have some process that takes a second, and it has a visual effect (animation). When the user presses some button we perform the effect. So we put these processes into promise queue. And now the user starts clicking that button very fast (with semantic "show me next result"). So with parameters `(concurrently: 1, maxQueue: 1)` we can easily skip some internal animations.

My fast, simplified and buggy proof-of-consept implementation: 

```
func applyPromise<T, U>(items: [T], task: (T -> Promise<U>), numberOfQueues: Int = 1) -> Promise<[U]> {

    return Promise<[U]> { fulfil, reject in
        var processed = 0
        var index = 0

        var results: [U] = []

        func block() {
            while index < items.count && (index - processed) < numberOfQueues {
                let promise = task(items[index])

                index += 1

                promise
                    .recover { error -> Promise<U> in
                        reject(error)
                        return Promise(error: error)
                    }
                    .then { result -> Void in
                        processed += 1
                        results.append(result)
                        block()
                    }
            }

            if index == items.count {
                fulfil(results)
            }
        }

        block()
    }
}
```

**Comment by shergin** [comment]
And the syntax with dedicated object `queue` allows working with streams of objects.

Oh, my syntax introduced in second comment is wrong! There is no `promiseResult`; every call of `addPromise()` should return new promise (like in mentioned npm library).

**Comment by mxcl** [comment]
> I believe, we actually cannot use/extend when because it accepts promises, so if we already have the promises, it immediately means that they already started resolving process (but we want to postpone it).

Yes I see, one of the features is that the promises are only generated on demand, which I like.

Also, I understand the `maxQueue` feature better now as well, and indeed: seems good.

Perhaps there is a better name than “queue” since it is concurrent and also debounces.

**Comment by shergin** [comment]
Should I do real implementation and pull-request?

**Comment by mxcl** [comment]
Please 👍🏻

**Comment by shergin** [comment]
Let's discuss the first approach. )
https://github.com/mxcl/PromiseKit/pull/445

(Of course, I will add tests later.)

**Comment by mxcl** [comment]
Merged: e50dc323c154f110a8defc819b916c09377e102e
